src/e3/anod/qualifiers_manager.py

```diff
@@ -0,0 +1,799 @@
+            not in new_key_value_qualifier["choices"]
+        ):
+            raise AnodError(
+                f"The {name} qualifier default value must be in choices. "
+                f'"{new_key_value_qualifier["default"]}" not in '
+                f'{new_key_value_qualifier["choices"]}'
+            )
+
+        self.qualifier_decls[name] = new_key_value_qualifier
+
+    def declare_component(
+        self,
+        name: str,
+        required_qualifier_configuration: dict[str, str],
+    ) -> None:
+        """Declare a new component.
+
+        A component is bound to a qualifier configuration (i.e. a set of qualifier
+        value). The provided component name is used if the qualifier meet the provided
+        qualifier values.
+
+        This method cannot be called after the end of the declaration phase.
+
+        :param name: A string representing the component name.
+        :param required_qualifier_configuration: The set (dictionary) of qualifiers
+            value corresponding to the build linked to the component.
+        """
+        if self.is_declaration_phase_finished:
+            raise AnodError(
+                "The component declaration is finished. It is not possible to declare a"
+                " new component after a name has been generated"
+            )
+
+        # Check name
+        valid_char = string.ascii_letters + string.digits + "-_."
+        if any(c not in valid_char for c in name):
+            raise AnodError(
+                f'The component name "{name}" contains an invalid character'
+            )
+
+        # Make sure the name has not been used yet
+        if name in self.raw_component_decls:
+            raise AnodError(f'The "{name}" component names is already used')
+
+        self.raw_component_decls[name] = required_qualifier_configuration
+
+    def __force_tag_qualifier_values(
+        self, qualifier_set: dict[str, str]
+    ) -> dict[str, str | bool]:
+        """Force the tag qualifiers values.
+
+        This function cannot be used before the end of the declaration phase.
+        :param qualifier_set: The qualifier dict for which the tag qualifiers values must
+        be set.
+        """
+        if not self.is_declaration_phase_finished:
+            raise AnodError(
+                "The tag qualifiers values cannot be added before the end of the "
+                "declaration phase"
+            )
+
+        qualifier_set_with_tag_values: dict[str, str | bool] = cast(
+            "dict[str, str | bool]", qualifier_set.copy()
+        )
+
+        for qualifier_name, qualifier in self.qualifier_decls.items():
+            if qualifier["type"] == self.__TAG_QUALIFIER:
+                qualifier_set_with_tag_values[qualifier_name] = (
+                    qualifier_name in qualifier_set
+                )
+
+        return qualifier_set_with_tag_values
+
+    def __force_default_values(self, qualifier_set: dict[str, str]) -> dict[str, str]:
```

Comment:
looking at the code I think it would be simpler to merge the two __force_* functions, your call